c/src/lib/libbsp/arm/shared/stm32fxxxx/can/hal-can.c

```diff
@@ -0,0 +1,664 @@
+ *                          .s2         - TQ in second bit segment
+ *                          .prescaler  - Preescaler
+ *                          .error      - Difference from requested baud
+ */
+CAN_Timing_Values rtems_can_get_timing_values
+(
+  uint32_t desired_baud
+)
+{
+
+  //TODO:: Determine best value.
+  //
+  // 1 BS = 1 Time Quata, or TQ.
+  //
+  // A TQ is the smallest unit of time that can be used to 
+  // create a CAN timing profile. It is Preescaler * (1 / Peripheral Clock)
+  //
+  // A CAN bit is divided up as follows:
+  //
+  //   +---+-------------+-------+
+  //   |SS |     BS1     | BS2   |
+  //   +---+-------------+-------+
+  //    ^  ^             ^       ^
+  //    |  |             |       |
+  //    |  \ Start Bit   |       \ End of Bit
+  //    \ Synq Seg      \ Sampling Point
+  //
+  //  There are also a Sync Segment that is 1 TQ
+  //
+  //
+  //  Therefore the baud rate is the 1/(1 TQ + BS1 + BS2)
+  //
+  //  Keep in mind that in calculations, we are storing BS1
+  //  and BS2 as multiples of 1TQ, a unitless quantity.
+  //
+  //  Keeping this in mind, the actual equation is
+  //
+  //  baud = 1 / (TQ * (1 + BS1 + BS2))
+  //
+
+  CAN_Timing_Values s_timeValues;
+
+  // We want to calculate the best
+  // possible values for BS1 BS2 and 
+  // the Prescalar. We are going to do 
+  // 2D optimization using a bruteforce
+  // algorithm, simply trying combinations
+  // and comparing them against our previous 
+  // best
+
+  float test_baud;
+
+  // Rather than optimizing for three variables,
+  // we are combining x = BS1 + BS2.
+  //
+  // Now our optimization is the best combination 
+  // of X and the Prescaler.
+  //
+  // We can calculate BS1 and BS2 later to 
+  // pick a good sampling point
+  int x;
+  int p;
+
+  float error = 1000000000;
+
+  // Store a very large error term
+  float best_error = 1000000000;
+  float best_x;
+  float best_p;
+
+  float clock_frequency = HAL_RCC_GetPCLK1Freq();
+  float tq = 1.0 / clock_frequency;
+
+  // x = S1 + S2
+  for (x = 9; x < 28; x++)
+  {
+    for (p = 0; p < 1025; p++)
+    {
+      test_baud = 1.0/ (tq * (x + 1) * (p) );
+
+      error = fabs(desired_baud- test_baud);
+      if (error < best_error) 
+      {
+        best_error = error;
+        best_x = x;
+        best_p = p;
+      }
+    }
+  }
+
+  // Calculate S1 and S2
+  // x = S1 + S2
+  int s1_plus_s2 = best_x;
+  int s1 = local_roundf(.875 * s1_plus_s2);
+  int s2 = s1_plus_s2 - s1;
+  s_timeValues.s1 = s1;
+  s_timeValues.s2 = s2;
+
+  s_timeValues.prescaler = best_p;
+  s_timeValues.error = best_error;
+  return s_timeValues;
+}
+
+
+
+
+void stm32_can_isr
+(
+  void * arg
+)
+{
+  HAL_CAN_IRQHandler((CAN_HandleTypeDef *) arg);
+}
+
+void stm32_can_rx0_isr
+(
+ void * arg
+)
+{
+  HAL_CAN_IRQHandler((CAN_HandleTypeDef *) arg);
+}
+
+void stm32_can_rx1_isr 
+( 
+  void * arg
+)
+{
+  HAL_CAN_IRQHandler((CAN_HandleTypeDef *) arg);
+}
+
+int stm32_can_set_filter
```

Comment:
The documenation will probably say CAN Receive ISR. I could maybe explain what the ISR does, but I don't know that much detail